### ISSUE TITLE
`jax.lax.map` is not compatible with `nnx.scan`

### ISSUE THREAD

**Issue by gaspardbb** [google/flax#5025]
### Problem description 

I'm running a model on a large batch of data, which can't be done in one-shot with `nnx.vmap`. Hence, I want to use `jax.lax.map` with a `batch_size` argument. 

While `jax.lax.map` works fine in general, it fails when a `nnx.scan` loop is involved in the `nnx.Module`.  

### System information
- OS Platform and Distribution: Mac OS 
- Flax, jax, jaxlib versions (obtain with `pip show flax jax jaxlib`:
```
Name: flax
Version: 0.12.0
---
Name: jax
Version: 0.7.2
---
Name: jaxlib
Version: 0.7.2
```
- Python version: `3.11.10`

### MRE 

```python
"""
Minimum Working Example: jax.lax.map with stateful NNX models

This script demonstrates that jax.lax.map works fine with stateful models
that don't use nnx.scan, but fails with models that use nnx.scan internally
due to trace level conflicts.
"""

import jax
import jax.numpy as jnp
from flax import nnx

# ==============================================================================
# Model 1: Stateful model WITHOUT nnx.scan
# ==============================================================================


class SimpleStatefulModel(nnx.Module):
    """A simple stateful model using dropout (no scan)."""

    def __init__(self, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)

    def __call__(self, x: jax.Array) -> jax.Array:
        """Forward pass."""
        h = self.linear1(x)
        h = nnx.relu(h)
        h = self.dropout(h, deterministic=True)  # Stateful but no scan
        h = self.linear2(h)
        return h.squeeze(-1)


# ==============================================================================
# Model 2: Stateful model WITH nnx.scan
# ==============================================================================


class StatefulModelWithScan(nnx.Module):
    """A stateful model that uses nnx.scan internally."""

    def __init__(self, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.rnn_cell = nnx.Linear(3, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)

    def cell(self, state: jax.Array, x: jax.Array) -> jax.Array:
        """RNN cell: update state based on input."""
        h = self.linear1(x)
        new_state = self.rnn_cell(state) + h
        new_state = nnx.relu(new_state)
        return new_state

    def __call__(self, x: jax.Array) -> jax.Array:
        """Forward pass with scan over sequence.

        Args:
            x: Input of shape (n_steps, 2)
        """

        @nnx.scan(
            in_axes=(
                nnx.StateAxes({...: nnx.Carry}),
                nnx.Carry,
                0,
            ),
            out_axes=(nnx.Carry, 1),
        )
        def unroll(model, state, x_step):
            new_state = model.cell(state, x_step)
            return new_state, None

        # Initial state
        initial_state = jnp.zeros(3)

        # Unroll over sequence
        final_state, _ = unroll(self, initial_state, x)

        # Apply final layers
        h = self.dropout(final_state, deterministic=True)
        output = self.linear2(h)
        return output.squeeze(-1)


# ==============================================================================
# Testing functions
# ==============================================================================


def test_with_vmap(model, data, model_name):
    """Test using nnx.vmap - should work for both models."""
    print(f'\n{"="*80}')
    print(f'Testing {model_name} with nnx.vmap')
    print('=' * 80)

    try:
        results = nnx.vmap(lambda m, x: m(x), in_axes=(None, 0))(model, data)
        print(f'✅ SUCCESS: Output shape: {results.shape}')
        print(f'   First 3 results: {results[:3]}')
        return True
    except Exception as e:
        print(f'❌ FAILED: {type(e).__name__}: {e}')
        return False


def test_with_lax_map(model, data, model_name):
    """Test using jax.lax.map with split/merge pattern."""
    print(f'\n{"="*80}')
    print(f'Testing {model_name} with jax.lax.map (split/merge pattern)')
    print('=' * 80)

    # Split model into graphdef and state
    graphdef, state = nnx.split(model)

    def pure_fn(x):
        """Pure function that merges model and calls it."""
        local_model = nnx.merge(graphdef, state)
        return local_model(x)

    try:
        results = jax.lax.map(pure_fn, data)
        print(f'✅ SUCCESS: Output shape: {results.shape}')
        print(f'   First 3 results: {results[:3]}')
        return True
    except Exception as e:
        print(f'❌ FAILED: {type(e).__name__}')
        print(f'   Error: {str(e)[:200]}...')
        return False


# ==============================================================================
# Main
# ==============================================================================


def main():
    rngs = nnx.Rngs(0)

    # =========================================================================
    # Test Model 1: Simple stateful model (no scan)
    # =========================================================================

    print('\n\n' + '#' * 80)
    print('# PART 1: Simple Stateful Model (no nnx.scan)')
    print('#' * 80)

    model1 = SimpleStatefulModel(rngs)
    data1 = jax.random.normal(jax.random.key(0), (8, 2))

    print('\nModel: SimpleStatefulModel')
    print(f'Input shape: {data1.shape}  (batch_size=8, features=2)')

    # Test with vmap
    vmap_works_1 = test_with_vmap(model1, data1, 'SimpleStatefulModel')

    # Test with lax.map
    lax_map_works_1 = test_with_lax_map(model1, data1, 'SimpleStatefulModel')

    # =========================================================================
    # Test Model 2: Stateful model with scan
    # =========================================================================

    print('\n\n' + '#' * 80)
    print('# PART 2: Stateful Model WITH nnx.scan')
    print('#' * 80)

    model2 = StatefulModelWithScan(rngs)
    # For this model, each input is a sequence of shape (n_steps, features)
    data2 = jax.random.normal(jax.random.key(1), (8, 5, 2))

    print('\nModel: StatefulModelWithScan')
    print(f'Input shape: {data2.shape}  (batch_size=8, n_steps=5, features=2)')
    print('Note: This model uses nnx.scan internally to process sequences')

    # Test with vmap
    vmap_works_2 = test_with_vmap(model2, data2, 'StatefulModelWithScan')

    # Test with lax.map - THIS WILL FAIL
    lax_map_works_2 = test_with_lax_map(model2, data2, 'StatefulModelWithScan')

    # =========================================================================
    # Summary
    # =========================================================================

    print('\n\n' + '=' * 80)
    print('SUMMARY')
    print('=' * 80)

    print('\nSimple Stateful Model (no nnx.scan):')
    print(f'  - nnx.vmap:    {"✅ WORKS" if vmap_works_1 else "❌ FAILS"}')
    print(f'  - jax.lax.map: {"✅ WORKS" if lax_map_works_1 else "❌ FAILS"}')

    print('\nStateful Model WITH nnx.scan:')
    print(f'  - nnx.vmap:    {"✅ WORKS" if vmap_works_2 else "❌ FAILS"}')
    print(f'  - jax.lax.map: {"✅ WORKS" if lax_map_works_2 else "❌ FAILS"}')


if __name__ == '__main__':
    main()
```

**Comment by vfdev-5** [comment]
@gaspardbb thanks for the report. I can make it work if we do not pass the model into `unroll`:
```python
class StatefulModelWithScan1(nnx.Module):
    """A stateful model that uses nnx.scan internally."""

    def __init__(self, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.rnn_cell = nnx.Linear(3, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)

    def cell(self, state: jax.Array, x: jax.Array) -> jax.Array:
        """RNN cell: update state based on input."""
        h = self.linear1(x)
        new_state = self.rnn_cell(state) + h
        new_state = nnx.relu(new_state)
        return new_state

    def __call__(self, x: jax.Array) -> jax.Array:
        @nnx.scan(
            in_axes=(
                nnx.Carry,  # <---- removed nnx.StateAxes
                0,
            ),
            out_axes=(nnx.Carry, 1),
        )
        def unroll(state, x_step):
            new_state = self.cell(state, x_step)  # <----- HERE
            return new_state, None

        # Initial state
        initial_state = jnp.zeros(3)

        # Unroll over sequence
        final_state, _ = unroll(initial_state, x)  # <----- HERE

        # Apply final layers
        h = self.dropout(final_state, deterministic=True)
        output = self.linear2(h)
        return output.squeeze(-1)
```

<details>
<summary>
Compare to vmap
</summary>

```python
import jax
import jax.numpy as jnp
from flax import nnx


class StatefulModelWithScan1(nnx.Module):
    """A stateful model that uses nnx.scan internally."""

    def __init__(self, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.rnn_cell = nnx.Linear(3, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)

    def cell(self, state: jax.Array, x: jax.Array) -> jax.Array:
        """RNN cell: update state based on input."""
        h = self.linear1(x)
        new_state = self.rnn_cell(state) + h
        new_state = nnx.relu(new_state)
        return new_state

    def __call__(self, x: jax.Array) -> jax.Array:
        @nnx.scan(
            in_axes=(
                nnx.Carry,
                0,
            ),
            out_axes=(nnx.Carry, 1),
        )
        def unroll(state, x_step):
            new_state = self.cell(state, x_step)
            return new_state, None

        # Initial state
        initial_state = jnp.zeros(3)

        # Unroll over sequence
        final_state, _ = unroll(initial_state, x)

        # Apply final layers
        h = self.dropout(final_state, deterministic=True)
        output = self.linear2(h)
        return output.squeeze(-1)


class StatefulModelWithScan2(nnx.Module):
    def __init__(self, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.rnn_cell = nnx.Linear(3, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)

    def cell(self, state: jax.Array, x: jax.Array) -> jax.Array:
        """RNN cell: update state based on input."""
        h = self.linear1(x)
        new_state = self.rnn_cell(state) + h
        new_state = nnx.relu(new_state)
        return new_state

    def __call__(self, x: jax.Array) -> jax.Array:
        @nnx.scan(
            in_axes=(
                nnx.StateAxes({...: nnx.Carry}),
                nnx.Carry,
                0,
            ),
            out_axes=(nnx.Carry, 1),
        )
        def unroll(model, state, x_step):
            new_state = model.cell(state, x_step)
            return new_state, None

        # Initial state
        initial_state = jnp.zeros(3)

        # Unroll over sequence
        final_state, _ = unroll(self, initial_state, x)

        # Apply final layers
        h = self.dropout(final_state, deterministic=True)
        output = self.linear2(h)
        return output.squeeze(-1)



def test_with_lax_map(model, data, model_name):
    graphdef, state = nnx.split(model)

    def pure_fn(x):
        """Pure function that merges model and calls it."""
        local_model = nnx.merge(graphdef, state)
        return local_model(x)

    return jax.lax.map(pure_fn, data)


def test_with_vmap(model, data, model_name):
    return nnx.vmap(lambda m, x: m(x), in_axes=(None, 0))(model, data)


def main():
    rngs = nnx.Rngs(0)
    model1 = StatefulModelWithScan1(rngs)
    rngs = nnx.Rngs(0)
    model2 = StatefulModelWithScan2(rngs)

    data2 = jax.random.normal(jax.random.key(1), (8, 5, 2))

    lax_map_works_2 = test_with_lax_map(model1, data2, 'StatefulModelWithScan1')
    vmap_works_2 = test_with_vmap(model2, data2, 'StatefulModelWithScan2')

    assert jax.numpy.allclose(lax_map_works_2, vmap_works_2)

if __name__ == '__main__':
    main()
```

</details>


I'll check in details what happens...

**Comment by gaspardbb** [comment]
Thanks for the feedback. 

I'm using [recurrent dropout](https://flax.readthedocs.io/en/latest/guides/randomness.html#recurrent-dropout). IIUC, in this case I do need to pass the model to `unroll` unfortunately.

**Comment by vfdev-5** [comment]
I'll take a look. To understand better, in your code, the dropout is outside the scanned function and applied once just before the final layer. Probably, `rnn_cell` contains it and you simplified it to Linear layer for this example?

**Comment by vfdev-5** [comment]
To fix trace level error, we can clone the model right before doing applying the unroll function:
```python
final_state, _ = unroll(nnx.clone(self), initial_state, x)
```

The same would be required if we use `jax.vmap` instead of `nnx.vmap`.

**Comment by gaspardbb** [comment]
> rnn_cell contains it and you simplified it to Linear layer for this example?

Yes! My code follows closely the `RNNCell` example.

> To fix trace level error, we can clone the model right before doing applying the unroll function:

Will that cause runtime/compile time overhead? 

Alternatively I'm implementing an nnx-variant of `jax.lax.map`, I will use that if there's overhead with `nnx.clone`

**Comment by vfdev-5** [comment]
> Will that cause runtime/compile time overhead?

Compiling and execution times with nnx transforms will be larger than with pure jax transforms even if using nnx.clone.
For example, nnx.vmap vs jax.vmap:
```
test_with_nnx_vmap: compile and execute once: 0.3286146344617009
test_with_jax_vmap: compile and execute once: 0.10536465514451265
test_with_nnx_vmap mean time: 0.099402 sec
test_with_nnx_vmap std dev  : 0.003971 sec
test_with_jax_vmap mean time: 0.094892 sec
test_with_jax_vmap std dev  : 0.001114 sec
```

<details>
<summary>
Code
</summary>

```python
import jax
import jax.numpy as jnp
from flax import nnx

import time
import timeit
import numpy as np
import statistics


class StatefulModelWithScan(nnx.Module):
    """A stateful model that uses nnx.scan internally."""

    def __init__(self, rngs: nnx.Rngs, do_clone: bool = False):
        self.linear1 = nnx.Linear(2, 3, rngs=rngs)
        self.rnn_cell = nnx.Linear(3, 3, rngs=rngs)
        self.dropout = nnx.Dropout(rate=0.1, rngs=rngs)
        self.linear2 = nnx.Linear(3, 1, rngs=rngs)
        self.do_clone = do_clone

    def cell(self, state: jax.Array, x: jax.Array) -> jax.Array:
        """RNN cell: update state based on input."""
        h = self.linear1(x)
        new_state = self.rnn_cell(state) + h
        new_state = nnx.relu(new_state)
        return new_state

    def __call__(self, x: jax.Array) -> jax.Array:
        @nnx.scan(
            in_axes=(
                nnx.StateAxes({...: nnx.Carry}),
                nnx.Carry,
                0,
            ),
            out_axes=(nnx.Carry, 1),
        )
        def unroll(model, state, x_step):
            new_state = model.cell(state, x_step)
            return new_state, None

        # Initial state
        initial_state = jnp.zeros(3)

        # Unroll over sequence
        mod = nnx.clone(self) if self.do_clone else self
        final_state, _ = unroll(mod, initial_state, x)

        # Apply final layers
        h = self.dropout(final_state, deterministic=True)
        output = self.linear2(h)
        return output.squeeze(-1)



def test_with_nnx_vmap(model, data):
    results = nnx.vmap(lambda m, x: m(x), in_axes=(None, 0))(model, data)
    return results


def test_with_jax_vmap(model, data):
    results = jax.vmap(lambda m, x: m(x), in_axes=(None, 0))(model, data)
    return results


def main():
    rngs = nnx.Rngs(0)
    model1 = StatefulModelWithScan(rngs)
    rngs = nnx.Rngs(0)
    model2 = StatefulModelWithScan(rngs, do_clone=True)
    data = jax.random.normal(jax.random.key(1), (8, 5, 2))

    start = time.perf_counter()
    test_with_nnx_vmap(model1, data).block_until_ready()
    elapsed = time.perf_counter() - start
    print("test_with_nnx_vmap: compile and execute once:", elapsed)

    start = time.perf_counter()
    test_with_jax_vmap(model2, data).block_until_ready()
    elapsed = time.perf_counter() - start
    print("test_with_jax_vmap: compile and execute once:", elapsed)

    n_repeat = 5   # How many times to repeat the timing
    n_number = 20    # How many times to run the function in each repeat
    times = timeit.repeat(
        lambda: test_with_nnx_vmap(model1, data).block_until_ready,
        repeat=n_repeat,
        number=n_number
    )
    normalized_times = np.array(times) / n_number
    print(f"test_with_nnx_vmap mean time: {statistics.mean(normalized_times):.6f} sec")
    print(f"test_with_nnx_vmap std dev  : {statistics.stdev(normalized_times):.6f} sec")

    times = timeit.repeat(
        lambda: test_with_jax_vmap(model2, data).block_until_ready,
        repeat=n_repeat,
        number=n_number
    )
    normalized_times = np.array(times) / n_number
    print(f"test_with_jax_vmap mean time: {statistics.mean(normalized_times):.6f} sec")
    print(f"test_with_jax_vmap std dev  : {statistics.stdev(normalized_times):.6f} sec")


if __name__ == '__main__':
    main()

```

</details>

**Comment by gaspardbb** [comment]
Indeed! The timing looks decent with `nnx.clone`. 

I ended up having my own `nnx.map` function, I'm ok to close the issue if needed or leave it open for bug tracking. Thanks for your help anyways!

**Comment by vfdev-5** [comment]
Let's close it. Feel free to reopen if needed. Thanks!
